File: Ahorn/entities/connectedBlocks/ConnectedSwapBlock.jl
module CommunalHelperConnectedSwapBlock

using ..Ahorn, Maple
using Ahorn.CommunalHelper

function swapFinalizer(entity)
    x, y = Ahorn.position(entity)
    width = Int(get(entity.data, "width", 8))

    entity.data["nodes"] = [(x + width, y)]
end

@mapdef Entity "CommunalHelper/ConnectedSwapBlock" ConnectedSwapBlock(
    x::Integer,
    y::Integer,
    width::Integer=Maple.defaultBlockWidth,
    height::Integer=Maple.defaultBlockWidth,
    theme::String="Normal",
    customGreenBlockTexture::String="",
    customRedBlockTexture::String="",
)

const placements = Ahorn.PlacementDict(
    "Connected Swap Block ($theme) (Communal Helper)" => Ahorn.EntityPlacement(
        ConnectedSwapBlock,
        "rectangle",
        Dict{String,Any}(
            "theme" => theme,
        ),
        swapFinalizer,
    ) for theme in Maple.swap_block_themes
)

placements["Connected Swap Block (Reskinnable) (Communal Helper)"] = Ahorn.EntityPlacement(
    ConnectedSwapBlock,
    "rectangle",
    Dict{String,Any}(
        "customGreenBlockTexture" => "CommunalHelper/customConnectedBlock/customConnectedBlock",
        "customRedBlockTexture" => "CommunalHelper/customConnectedBlock/customConnectedBlock",
    ),
    swapFinalizer,
)

Ahorn.editingOptions(entity::ConnectedSwapBlock) = Dict{String,Any}(
    "theme" => Maple.swap_block_themes,
)

Ahorn.nodeLimits(entity::ConnectedSwapBlock) = 1, 1

Ahorn.minimumSize(entity::ConnectedSwapBlock) = 16, 16
Ahorn.resizable(entity::ConnectedSwapBlock) = true, true

function Ahorn.selection(entity::ConnectedSwapBlock)
    x, y = Ahorn.position(entity)
    stopX, stopY = Int.(entity.data["nodes"][1])

    width = Int(get(entity.data, "width", 8))
    height = Int(get(entity.data, "height", 8))

    return [
        Ahorn.Rectangle(x, y, width, height),
        Ahorn.Rectangle(stopX, stopY, width, height),
    ]
end

function getTextures(entity::ConnectedSwapBlock)
    theme = lowercase(get(entity, "theme", "normal"))
    themePath = (theme == "normal") ? "" : string(theme, "/")

    return (
        "objects/swapblock/$(themePath)blockRed",
        "objects/swapblock/$(themePath)target",
        "objects/swapblock/$(themePath)midBlockRed00",
        "objects/CommunalHelper/connectedSwapBlock/$(themePath)innerCornersRed",
    )
end

function renderTrail(ctx, x::Number, y::Number, width::Number, height::Number, trail::String)
    tilesWidth = div(width, 8)
    tilesHeight = div(height, 8)

    for i in 2:tilesWidth - 1
        Ahorn.drawImage(ctx, trail, x + (i - 1) * 8, y + 2, 6, 0, 8, 6)
        Ahorn.drawImage(ctx, trail, x + (i - 1) * 8, y + height - 8, 6, 14, 8, 6)
    end

    for i in 2:tilesHeight - 1
        Ahorn.drawImage(ctx, trail, x + 2, y + (i - 1) * 8, 0, 6, 6, 8)
        Ahorn.drawImage(ctx, trail, x + width - 8, y + (i - 1) * 8, 14, 6, 6, 8)
    end

    for i in 2:tilesWidth - 1, j in 2:tilesHeight - 1
        Ahorn.drawImage(ctx, trail, x + (i - 1) * 8, y + (j - 1) * 8, 6, 6, 8, 8)
    end

    Ahorn.drawImage(ctx, trail, x + width - 8, y + 2, 14, 0, 6, 6)
    Ahorn.drawImage(ctx, trail, x + width - 8, y + height - 8, 14, 14, 6, 6)
    Ahorn.drawImage(ctx, trail, x + 2, y + 2, 0, 0, 6, 6)
    Ahorn.drawImage(ctx, trail, x + 2, y + height - 8, 0, 14, 6, 6)
end

function renderSwapBlock(ctx::Ahorn.Cairo.CairoContext, x::Number, y::Number, width::Number, height::Number, midResource::String, block::String, innerCorners::String, txOffset::Integer, entity::ConnectedSwapBlock, room::Maple.Room)
    midSprite = Ahorn.getSprite(midResource, "Gameplay")

    tileWidth = div(width, 8)
    tileHeight = div(height, 8)

    rects = getExtensionRectangles(room)
    rx, ry = Ahorn.position(entity)
    rect = Ahorn.Rectangle(rx, ry, width, height)

    if !(rect in rects)
        push!(rects, rect)
    end

    for i in 1:tileWidth, j in 1:tileHeight
        drawX, drawY = (i - 1) * 8, (j - 1) * 8

        closedLeft = !notAdjacent(entity, drawX - 8, drawY, rects)
        closedRight = !notAdjacent(entity, drawX + 8, drawY, rects)
        closedUp = !notAdjacent(entity, drawX, drawY - 8, rects)
        closedDown = !notAdjacent(entity, drawX, drawY + 8, rects)
        completelyClosed = closedLeft && closedRight && closedUp && closedDown

        if completelyClosed
            if notAdjacent(entity, drawX + 8, drawY - 8, rects)
                # up right
                Ahorn.drawImage(ctx, innerCorners, x + drawX, y + drawY, 8 + txOffset, 0, 8, 8)

            elseif notAdjacent(entity, drawX - 8, drawY - 8, rects)
                # up left
                Ahorn.drawImage(ctx, innerCorners, x + drawX, y + drawY, 0 + txOffset, 0, 8, 8)

            elseif notAdjacent(entity, drawX + 8, drawY + 8, rects)
                # down right
                Ahorn.drawImage(ctx, innerCorners, x + drawX, y + drawY, 8 + txOffset, 8, 8, 8)

            elseif notAdjacent(entity, drawX - 8, drawY + 8, rects)
                # down left
                Ahorn.drawImage(ctx, innerCorners, x + drawX, y + drawY, 0 + txOffset, 8, 8, 8)

            else
                # entirely surrounded, fill tile
                Ahorn.drawImage(ctx, block, x + drawX, y + drawY, 8, 8, 8, 8)
            end
        else
            if closedLeft && closedRight && !closedUp && closedDown
                Ahorn.drawImage(ctx, block, x + drawX, y + drawY, 8, 0, 8, 8)

            elseif closedLeft && closedRight && closedUp && !closedDown
                Ahorn.drawImage(ctx, block, x + drawX, y + drawY, 8, 16, 8, 8)

            elseif closedLeft && !closedRight && closedUp && closedDown
                Ahorn.drawImage(ctx, block, x + drawX, y + drawY, 16, 8, 8, 8)

            elseif !closedLeft && closedRight && closedUp && closedDown
                Ahorn.drawImage(ctx, block, x + drawX, y + drawY, 0, 8, 8, 8)

            elseif closedLeft && !closedRight && !closedUp && closedDown
                Ahorn.drawImage(ctx, block, x + drawX, y + drawY, 16, 0, 8, 8)

            elseif !closedLeft && closedRight && !closedUp && closedDown
                Ahorn.drawImage(ctx, block, x + drawX, y + drawY, 0, 0, 8, 8)

            elseif !closedLeft && closedRight && closedUp && !closedDown
                Ahorn.drawImage(ctx, block, x + drawX, y + drawY, 0, 16, 8, 8)

            elseif closedLeft && !closedRight && closedUp && !closedDown
                Ahorn.drawImage(ctx, block, x + drawX, y + drawY, 16, 16, 8, 8)
            end
        end
    end

    Ahorn.drawImage(ctx, midSprite, x + div(width - midSprite.width, 2), y + div(height - midSprite.height, 2))
end

function Ahorn.renderSelectedAbs(ctx::Ahorn.Cairo.CairoContext, entity::ConnectedSwapBlock, room::Maple.Room)
    startX, startY = Int(entity.data["x"]), Int(entity.data["y"])
    stopX, stopY = Int.(entity.data["nodes"][1])

    width = Int(get(entity.data, "width", 32))
    height = Int(get(entity.data, "height", 32))

    frame, _, mid, innerCorners = getTextures(entity)
    customBlockTexture = String(get(entity.data, "customRedBlockTexture", ""))
    hasCustomTexture = customBlockTexture != ""
    txOffset = 0

    if hasCustomTexture 
        frame = innerCorners = "objects/" * customBlockTexture
        txOffset = 24
    end

    renderSwapBlock(ctx, stopX, stopY, width, height, mid, frame, innerCorners, txOffset, entity, room)
    Ahorn.drawArrow(ctx, startX + width / 2, startY + height / 2, stopX + width / 2, stopY + height / 2, Ahorn.colors.selection_selected_fc, headLength=6)
end

function Ahorn.renderAbs(ctx::Ahorn.Cairo.CairoContext, entity::ConnectedSwapBlock, room::Maple.Room)
    startX, startY = Int(entity.data["x"]), Int(entity.data["y"])
    stopX, stopY = Int.(entity.data["nodes"][1])

    width = Int(get(entity.data, "width", 32))
    height = Int(get(entity.data, "height", 32))

    frame, trail, mid, innerCorners = getTextures(entity)
    customBlockTexture = String(get(entity.data, "customRedBlockTexture", ""))
    hasCustomTexture = customBlockTexture != ""
    txOffset = 0

    if hasCustomTexture 
        frame = innerCorners = "objects/" * customBlockTexture
        txOffset = 24
    end

    renderTrail(ctx, min(startX, stopX), min(startY, stopY), abs(startX - stopX) + width, abs(startY - stopY) + height, trail)
    renderSwapBlock(ctx, startX, startY, width, height, mid, frame, innerCorners, txOffset, entity, room)
end

end

File: Loenn/entities/connected_blocks/connected_swap_block.lua
local connectedEntities = require("helpers.connected_entities")
local drawableSprite = require("structs.drawable_sprite")
local utils = require("utils")
local enums = require("consts.celeste_enums")
local communalHelper = require("mods").requireFromPlugin("libraries.communal_helper")

local connectedSwapBlock = {}

connectedSwapBlock.name = "CommunalHelper/ConnectedSwapBlock"
connectedSwapBlock.depth = -9999
connectedSwapBlock.minimumSize = {16, 16}
connectedSwapBlock.nodeLimits = {1, 1}
connectedSwapBlock.fieldInformation = {
    theme = {
        editable = false,
        options = enums.swap_block_themes
    }
}

connectedSwapBlock.placements = {
    {
        name = "normal",
        placementType = "rectangle",
        data = {
            width = 16,
            height = 16,
            theme = "Normal",
            customGreenBlockTexture = "",
            customRedBlockTexture = "",
        }
    },
    {
        name = "reskinnable",
        placementType = "rectangle",
        data = {
            width = 16,
            height = 16,
            theme = "Normal",
            customGreenBlockTexture = "CommunalHelper/customConnectedBlock/customConnectedBlock",
            customRedBlockTexture = "CommunalHelper/customConnectedBlock/customConnectedBlock",
        }
    }
}

local function getSearchPredicate()
    return function(target)
        return target._name == "CommunalHelper/SolidExtension"
    end
end

local function getTileSprite(entity, x, y, block, inner, txo, rectangles)
    local hasAdjacent = connectedEntities.hasAdjacent

    local drawX, drawY = (x - 1) * 8, (y - 1) * 8

    local closedLeft = hasAdjacent(entity, drawX - 8, drawY, rectangles)
    local closedRight = hasAdjacent(entity, drawX + 8, drawY, rectangles)
    local closedUp = hasAdjacent(entity, drawX, drawY - 8, rectangles)
    local closedDown = hasAdjacent(entity, drawX, drawY + 8, rectangles)
    local completelyClosed = closedLeft and closedRight and closedUp and closedDown

    local quadX, quadY = false, false
    local frame = block

    if completelyClosed then
        frame = inner
        if not hasAdjacent(entity, drawX + 8, drawY - 8, rectangles) then
            quadX, quadY = 8 + txo, 0
        elseif not hasAdjacent(entity, drawX - 8, drawY - 8, rectangles) then
            quadX, quadY = 0 + txo, 0
        elseif not hasAdjacent(entity, drawX + 8, drawY + 8, rectangles) then
            quadX, quadY = 8 + txo, 8
        elseif not hasAdjacent(entity, drawX - 8, drawY + 8, rectangles) then
            quadX, quadY = 0 + txo, 8
        else
            quadX, quadY = 8, 8
            frame = block
        end
    else
        if closedLeft and closedRight and not closedUp and closedDown then
            quadX, quadY = 8, 0

        elseif closedLeft and closedRight and closedUp and not closedDown then
            quadX, quadY = 8, 16

        elseif closedLeft and not closedRight and closedUp and closedDown then
            quadX, quadY = 16, 8

        elseif not closedLeft and closedRight and closedUp and closedDown then
            quadX, quadY = 0, 8

        elseif closedLeft and not closedRight and not closedUp and closedDown then
            quadX, quadY = 16, 0

        elseif not closedLeft and closedRight and not closedUp and closedDown then
            quadX, quadY = 0, 0

        elseif not closedLeft and closedRight and closedUp and not closedDown then
            quadX, quadY = 0, 16

        elseif closedLeft and not closedRight and closedUp and not closedDown then
            quadX, quadY = 16, 16
        end
    end

    if quadX and quadY then
        local sprite = drawableSprite.fromTexture(frame, entity)

        sprite:addPosition(drawX, drawY)
        sprite:useRelativeQuad(quadX, quadY, 8, 8)

        return sprite
    end
end

local function getConnectedSwapBlockThemeData(entity)
    local theme = string.lower(entity.theme or "normal")
    local themePath = (theme == "normal") and "" or (theme .. "/")

    local path = "objects/swapblock/" .. themePath .. "target"
    local mid = "objects/swapblock/" .. themePath .. "midBlockRed00"

    local customBlockTexture = entity.customRedBlockTexture or ""
    if customBlockTexture ~= "" then
        local full = "objects/" .. customBlockTexture
        return {
            block = full,
            inner = full,
            path = path,
            mid = mid,
            txOffset = 24,
        }
    end

    return {
        block = "objects/swapblock/" .. themePath .. "blockRed",
        inner = "objects/CommunalHelper/connectedSwapBlock/" .. themePath .. "innerCornersRed",
        path = path,
        mid = mid,
        txOffset = 0,
    }
end

local trailNinePatchOptions = {
    mode = "fill",
    borderMode = "repeat",
    useRealSize = true
}

local function addBlockSprites(sprites, room, entity, w, h, tw, th, themeData)
    local relevantBlocks = utils.filter(getSearchPredicate(), room.entities)
    connectedEntities.appendIfMissing(relevantBlocks, entity)
    local rectangles = connectedEntities.getEntityRectangles(relevantBlocks)

    for i = 1, tw do
        for j = 1, th do
            local sprite = getTileSprite(entity, i, j, themeData.block, themeData.inner, themeData.txOffset, rectangles)

            if sprite then
                table.insert(sprites, sprite)
            end
        end
    end

    local middleSprite = drawableSprite.fromTexture(themeData.mid, entity)
    middleSprite:addPosition(math.floor(w / 2), math.floor(h / 2))
    middleSprite.depth = -9999
    table.insert(sprites, middleSprite)
end

function connectedSwapBlock.sprite(room, entity)
    local x, y = entity.x or 0, entity.y or 0
    local width, height = entity.width or 16, entity.height or 16
    local nodes = entity.nodes or {}
    local nodeX, nodeY = nodes[1].x or x, nodes[1].y or y
    local tileWidth, tileHeight = math.ceil(width / 8), math.ceil(height / 8)

    local themeData = getConnectedSwapBlockThemeData(entity)

    local sprites = {}

    communalHelper.addTrailSprites(sprites, x, y, nodeX, nodeY, width, height, themeData.path)
    addBlockSprites(sprites, room, entity, width, height, tileWidth, tileHeight, themeData)

    return sprites
end

function connectedSwapBlock.nodeSprite(room, entity)
    local x, y = entity.x or 0, entity.y or 0
    local width, height = entity.width or 16, entity.height or 16
    local nodes = entity.nodes or {}
    local nodeX, nodeY = nodes[1].x or x, nodes[1].y or y
    local tileWidth, tileHeight = math.ceil(width / 8), math.ceil(height / 8)

    local themeData = getConnectedSwapBlockThemeData(entity)

    local sprites = {}

    addBlockSprites(sprites, room, entity, width, height, tileWidth, tileHeight, themeData)

    for _, sprite in ipairs(sprites) do
        sprite:addPosition(nodeX - x, nodeY - y)
    end

    return sprites
end

function connectedSwapBlock.selection(room, entity)
    local nodes = entity.nodes or {}
    local x, y = entity.x or 0, entity.y or 0
    local nodeX, nodeY = nodes[1].x or x, nodes[1].y or y
    local width, height = entity.width or 8, entity.height or 8

    return utils.rectangle(x, y, width, height), {utils.rectangle(nodeX, nodeY, width, height)}
end

return connectedSwapBlock
